### DISCUSSION TITLE
使用自定义 Module 解决 Objective-C 库的引用 | Alex Lin's Notes

### DISCUSSION THREAD

**Discussion by xwal** [chaoskyme/chaoskyme.github.io#322] [Comment]
https://chaosky.tech/2020/08/03/cocoapods-custom-swift-module/ 

 LLVM Module A module is a single unit of code distribution—a framework or application that is built and shipped as a single unit and that can be imported by another module with Swift’s import keyword.

**Comment by no5ugar** [comment]
请问有demo吗？ 我根据步骤尝试但是还是找不到module

**Comment by xwal** [comment]
用这个demo试试，https://github.com/chaoskyx/Demo/raw/master/GDTPackage.zip

**Comment by no5ugar** [comment]
> @chaoskyx
> 用这个demo试试，https://github.com/chaoskyx/Demo/raw/master/GDTPackage.zip

谢谢，我看了下，demo解决的是swift的组件里面引用了oc的.a静态库的情况如何适配（很不错的方案，赞），但其实我想了解的是怎么手动在oc(非.a静态库)组件里面创建modulemap并支持在主工程引入，不通过 cocoapods 在podspec里面设置{'DEFINES_MODULE' => 'YES'} 或者 podfile 里面设置 modular_headers等方式实现，因为我看到 podspec 是支持 module_map 的配置的 "spec.module_map = 'source/module.modulemap'" 但是设置后在主工程并找不到这个module

**Comment by xwal** [comment]
> 谢谢，我看了下，demo解决的是swift的组件里面引用了oc的.a静态库的情况如何适配（很不错的方案，赞），但其实我想了解的是怎么手动在oc(非.a静态库)组件里面创建modulemap并支持在主工程引入，不通过 cocoapods 在podspec里面设置{'DEFINES_MODULE' => 'YES'} 或者 podfile 里面设置 modular_headers等方式实现，因为我看到 podspec 是支持 module_map 的配置的 "spec.module_map = 'source/module.modulemap'" 但是设置后在主工程并找不到这个module

在framework 的Build Settings 添加以下内容，并且保证modulemap的路径正确试试

```
DEFINES_MODULE = YES; 
MODULEMAP_FILE = "Target Support Files/GDTPackage/GDTPackage.modulemap";
PRODUCT_MODULE_NAME = GDTPackage;
```

**Comment by no5ugar** [comment]
是我误解了，我以为用了{'DEFINES_MODULE' => 'YES'}，cocoapods会自己生成modulemap，但仔细看了下，用了{'DEFINES_MODULE' => 'YES'}是使用我自定义的modulemap，那可以解决我的问题了，谢谢

**Comment by cocoongg** [comment]
你好，我正在弄OC和Swift混编，并且组件化，将各个组件打成framework的形式，请问能不能加练习方式交流一下，感谢
